rfc/drafts/indexer-ideas-v1.md

```diff
@@ -0,0 +1,303 @@
+- POST `/distance`   - Calculate metric-space distance between 2 images.
+- POST `/dupe_check` - Return calibrated-probability that a pair of images are duplicates.
+- GET  `/ping`       - Uptime check.
+- POST `/record_dupe`      - User labels a set of images as dupes / non-dupes. This is added to training data.
+- POST `/record_relevancy` - User labels the relevancy of a set of search results for a query. This is added to training data.
+
+
+TODO:
+
+- Fill in exact details of JSON parameters for above endpoints.
+- `/dupe_add` and `/relevancy_rate` - these are training feedback endpoints. Other types of training data and hyper-parameter settings will be also fed into the system elsewhere, but these two endpoints are included here because they may be end-user-facing?
+-   Perhaps these should be an immediate feedback tool, whose input supersedes that from the slower index re-training process?
+
+
+## SEARCH / DEDUPE ROADMAP:
+
+Version 1:
+
+  - Centralized.
+  - Off-the-shelf components.
+  - No focus on provider incentives.
+  - Unsupervised / distantly-supervised models.
+  - No custom relevancy / dedupe training.
+  - Skipping some components in V1, e.g. the "high-accuracy-model". See below.
+  - Generic wrapping APIs for each component, so they can be easily swapped out for others.
+
+Long-term:
+
+  - Distributed across machines.
+  - Distributed across users. Anyone can run an indexing node helps serve someone else's index(?)
+  - More attention to optimizing incentives among resource-providers / service-providers.
+  - Custom-trained for these particular relevancy / dedupe tasks & datasets.
+  - Accepting explicit user feedback signals(?)
+  - Mitigating attacks: resource exhaustion, bad feedback, high-trust actors gone bad, etc.
+  - Other possible ideas, see below.
+
+
+## SCALING ANN INDEXES:
+
+Distributed computing for approximate nearest neighbor (ANN) indexes requires some special consideration.
+
+Sharding:
+
+- Typical databases you heavily pre-shard to support future scaling, e.g. pre-sharded into 65000 shards.
+- Big efficiency penalties for sharding of ANN indexes. Generally minimal-possible sharding per node is optimal.
+
+- Balance between pre-sharding penalties and re-indexing penalties?
+- Consideration needed.
+
+GPU advances make smaller clusters, or even single-machine information retrieval go much further:
+
+   - See the recent e.g. CVPR paper on GPU-based ANN lookup.
+   - 1 GPU = ~300x CPU cores.
+   - When immediate results are required, limited by dataset size due to GPU RAM.
+   - Only worry about failure of a single machine, instead of many. Reduced 95th-percentile worse-case response times?
+
+Typical distributed approach is something like the "shards & replicas" approach of ElasticSearch / Lucene:
+
+   - Each shard sees a subset of the data.
+   - Each replica set contains enough shards to fit entire dataset.
```

Comment:
Yes, was thinking to use ES and its supporting tools for ANN lookup / text indexing in V1.

Seems there would be a smooth iteration path, from a simple ES-backed solution, where ES does most of the work -> swapping in custom vectorization and re-ranking models, with ES just being used for ANN indexing and lookup -> finally swapping out ES for something more custom suited for the P2P role.

Could maybe consider another step in this iteration, where each distributed node would hold the entire search index. I'll run a quick benchmark to get a feel for the RAM it takes for 1 indexing node to handle e.g. 1MM, 50MM, 1B images. MMAPing large indexes is also possible, but I'd approach that cautiously for time-sensitive queries.